sdk/python/kfp/compiler/_container_builder.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import tarfile
+import tempfile
+import os
+import uuid
+
+class ContainerBuilder(object):
+  """
+  ContainerBuilder helps build a container image
+  """
+  def __init__(self, gcs_staging):
+    if not self._check_gcs_path(gcs_staging):
+      raise Exception('ContainerBuilder __init__ failure.')
+    self._gcs_staging = gcs_staging
+
+  def _check_gcs_path(self, gcs_path):
+    """ _check_gcs_path check both the path validity and write permissions """
+    logging.info('Checking path: {}...'.format(gcs_path))
+    if not gcs_path.startswith('gs://'):
+      logging.error('Error: {} should be a GCS path.'.format(gcs_path))
+      return False
+    return True
+
+  def _generate_kaniko_spec(self, namespace, context, docker_filename, target_image):
+    """_generate_kaniko_yaml generates kaniko job yaml based on a template yaml """
+    content = {
+        'apiVersion': 'v1',
+        'metadata': {
+            'generateName': 'kaniko-',
+            'namespace': namespace,
+        },
+        'kind': 'Pod',
+        'spec': {
+            'restartPolicy': 'Never',
+            'containers': [{
+                'name': 'kaniko',
+                'args': ['--cache=true',
+                         '--dockerfile=' + docker_filename,
+                         '--context=' + context,
+                         '--destination=' + target_image],
+                'image': 'gcr.io/kaniko-project/executor@sha256:78d44ec4e9cb5545d7f85c1924695c89503ded86a59f92c7ae658afa3cff5400',
+                'env': [{
+                    'name': 'GOOGLE_APPLICATION_CREDENTIALS',
+                    'value': '/secret/gcp-credentials/user-gcp-sa.json'
+                }],
+                'volumeMounts': [{
+                    'mountPath': '/secret/gcp-credentials',
+                    'name': 'gcp-credentials',
+                }],
+            }],
+            'volumes': [{
+                'name': 'gcp-credentials',
+                'secret': {
+                    'secretName': 'user-gcp-sa',
+                },
+            }],
+            'serviceAccountName': 'default'}
+    }
+    return content
+
+  def _wrap_dir_in_tarball(self, tarball_path, dir_name):
+    """ _wrap_files_in_tarball creates a tarball for all the files in the directory"""
+    old_wd = os.getcwd()
+    os.chdir(dir_name)
```

Comment:
Curious why this is needed? Does tarbar lib relies on cwd to find file?

Reply:
Good question. The kaniko build requires the tarball to contain only the files. In other words, it will fail if the context tarball contain a directory that contains all the build files. 
'chdir' is needed here to make sure that that tarball contains the files with no intermediate directories created.

Reply:
`tarfile.add` supports adding the whole directory at once and it also supports specifying the name for it in the archive:
The function can be simplified to
```python
with tarfile.open(tarball_path, 'w:gz') as tarball:
    tarball.add(dir_name, arcname=".")
```

Reply:
I was hoping that it worked. However, this will create an intermediate directory called '.'

Reply:
> I was hoping that it worked. However, this will create an intermediate directory called '.'

You can also specify an empty string there. `arcname=""`

Reply:
https://github.com/kubeflow/pipelines/pull/1703

Reply:
Let's see if the sample can pass since it adds an empty tarfile object in the tarball. However, it does not add the intermediate directory, which is good.